client/db/src/storage_cache.rs

```diff
@@ -343,11 +345,29 @@ impl<B: BlockT> CacheChanges<B> {
 		);
 		let cache = &mut *cache;
 		// Filter out committing block if any.
-		let enacted: Vec<_> = enacted
+		let mut enacted: Vec<_> = enacted
 			.iter()
 			.filter(|h| commit_hash.as_ref().map_or(true, |p| *h != p))
 			.cloned()
 			.collect();
+
+		let mut retracted_vec;
+		let mut retracted = retracted;
+		if let Some(commit_hash) = &commit_hash {
+			if let Some(m) = cache.modifications.iter_mut().find(|m| &m.hash == commit_hash) {
+				if m.is_canon != is_best {
+					// Same block comitted twice with different state changes.
+					// Treat it as reenacted/retracted.
+					if is_best {
+						enacted.push(commit_hash.clone());
+					} else {
+						retracted_vec = retracted.to_vec();
```

Comment:
Just a question about this. This would only happen when we import the block a second time with `best = false` and before we had imported it with `best = true`?

Reply:
Right. The opposite case (`best = true` followed by `best = false`) is also handled though